File: object-oriented-programming/virtual-world-java/src/Game/Model/World.java
package Game.Model;

import Game.Model.Animals.Human;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.function.Consumer;

public class World implements Serializable {

    private static final int MAX_ORGANISMS = 6400;
    private static final int LOG_SIZE = 50;
    private final Organism[][] grid;
    private final int width;
    private final int height;
    private int round = 0;
    private int lastKey = 0;
    private final List<Organism> organisms = new ArrayList<>();
    private final Deque<String> history = new ArrayDeque<>(LOG_SIZE);
    private transient Consumer<String> listener;

    public World(int width, int height){
        this.width = width;
        this.height = height;
        this.grid = new Organism[height][width];
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public int getRound() {
        return round;
    }

    public boolean inBounds(int y, int x){
        return (y >= 0 && y < height && x >= 0 && x < width);
    }

    public boolean isFree(int y, int x){
        return (inBounds(y, x) && grid[y][x] == null);
    }

    public Organism getOrganism(int y, int x){
        if(!inBounds(y, x)) return null;
        return grid[y][x];
    }

    public void setOrganism(Organism organism, int y, int x){
        if(inBounds(y, x)){
            grid[y][x] = organism;
        }
    }

    public void addOrganism(Organism organism) {
        if (organisms.size() < MAX_ORGANISMS) {
            organisms.add(organism);
        }
    }

    public void makeTurn(){
        round++;
        organisms.sort(Comparator.comparingInt(Organism::getInitiative).reversed()
                .thenComparingInt(Organism::getAge).reversed());

        for(Organism organism : new ArrayList<>(organisms)){
            if(!organism.isDead()){
                organism.action();
                organism.incrementAge();
            }
        }
        organisms.removeIf(organism -> organism.isDead() && organism.getSpecies() != Species.HUMAN);
    }

    public void setLastKey(int k) {
        lastKey = k;
    }

    public final int getLastKey() {
        return lastKey;
    }

    public void saveToFile(Path file) throws IOException {
        try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(file))) {
            oos.writeObject(this);
        }
    }

    public static World loadFromFile(Path file) throws IOException, ClassNotFoundException {
        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(file))) {
            return (World) ois.readObject();
        }
    }

    public void setReportListener(Consumer<String> listener){
        this.listener = listener;
    }

    public void report(String msg) {
        if (history.size() == LOG_SIZE) history.removeFirst();
        history.addLast(msg);
        if (listener != null) listener.accept(msg);
    }

    public List<String> getHistorySnapshot() {
        return new ArrayList<>(history);
    }

    public Human assignHuman() {
        for (Organism organism : organisms) {
            if (organism instanceof Human) return (Human)organism;
        }
        return null;
    }
}

File: object-oriented-programming/virtual-world-python/world.py
from collections import deque
import pickle
from species import Species

class World :
    LOG_SIZE = 50

    def __init__(self, width, height) :
        self._width = width
        self._height = height
        self.grid = [[None]*height for _ in range(width)]
        self.organisms = []
        self._turn = 0
        self.history = deque(maxlen = self.LOG_SIZE)
        self.listener = None
        self._last_key = 0

    @property
    def width(self) :
        return self._width
    
    @property
    def height(self) :
        return self._height
    
    @property
    def turn(self) :
        return self._turn
    
    def in_bounds(self, x, y) :
        return (0 <= x < self._width and 0 <= y < self._height)
    
    def is_free(self, x, y) :
        return (self.in_bounds(x,y) and self.grid[x][y] is None)
    
    def get_organism(self, x, y) :
        return (self.grid[x][y] if self.in_bounds(x, y) else None)
    
    def set_organism(self, organism, x, y) :
        self.grid[x][y] = organism

    def add_organism(self, organism) :
        self.organisms.append(organism)

    def report(self, msg: str):
        self.history.append(msg)
        if self.listener: self.listener(msg)

    def add_listener(self, fn):
        self.listener = fn

    def get_history(self):
        return list(self.history)

    def make_turn(self):
        self._turn += 1
        for org in list(self.organisms):
            if not org.is_dead():
                org.action()
                org.increment_age()

        survivors = []
        for org in self.organisms:
            if org.is_dead():
                if self.get_organism(org.x, org.y) is org:
                    self.set_organism(None, org.x, org.y)
            else:
                survivors.append(org)
        self.organisms = survivors


    def set_last_key(self, key) :
        self._last_key = key

    def get_last_key(self) :
        return self._last_key
    
    def save(self, path):
            saved_listener = self.listener
            self.listener = None

            with open(path, "wb") as f:
                pickle.dump(self, f)

            self.listener = saved_listener


    @classmethod
    def load(cls, path) :
        with open(path, "rb") as file:
            world = pickle.load(file)

        world.listener = None
        return world
    
    def assign_human(self):
        for org in self.organisms:
            if org.species == Species.HUMAN:
                return org
        raise RuntimeError("No human found in loaded world!")

